### PR TITLE
Add delete endpoint

### PR DESCRIPTION
**Pull request by Igoraamc**
What does this implement/fix? Explain your changes.
---------------------------------------------------
This pull request implements the delete feature.

- Add delete data from DeviceID
- Add delete all collection 😭

Does this close any currently open issues?
------------------------------------------
Any relevant logs, error output, etc?
-------------------------------------
Any other comments?
-------------------
Where has this been tested?
---------------------------

**Operating System/Platform:** MACOS Catalina

**GO Version:** go 1.13.7 darwin/amd64

### CODE REVIEW

File: pkg/controllers/data.go [hunk 1]
```diff
@@ -89,6 +89,27 @@ func (d *DataController) Create(w http.ResponseWriter, r *http.Request) {
 
 }
 
+func (d *DataController) DeleteAll(w http.ResponseWriter, r *http.Request) {
```
Comment (netoax) [inline pkg/controllers/data.go]: Please, break this commit in two in order to add `DeleteAll` and `DeleteByDevice ` separately.

File: pkg/interactor/data.go [hunk 2]
```diff
@@ -44,6 +44,16 @@ func (d *DataInteractor) Create(data Data) error {
 	return err
 }
 
+func (d *DataInteractor) DeleteByDevice(deviceId string) (error error) {
```
Comment (netoax) [inline pkg/interactor/data.go]: It's advisable to create one interactor file for each use case to avoid mixing up the business logic. In this case, you should create another file named as `delete_data.go`.  One possibility is to create a DeleteData interactor (with only one method to be externally called - `DeleteData()`) that receives the `deviceId` option so that it can decide to remove all or by `deviceId`.

File: pkg/server/server.go [hunk 3]
```diff
@@ -66,9 +66,11 @@ func (s *Server) createRouters() *negroni.Negroni {
 
 	n.UseHandler(r)
 
-	r.HandleFunc("/data/{deviceId}", dataController.GetAll).Methods("GET")
-	r.HandleFunc("/data/{deviceId}/sensor/{id}", dataController.GetByID).Methods("GET")
+	r.HandleFunc("/data/{user}", dataController.GetAll).Methods("GET")
```
Comment (netoax) [inline pkg/server/server.go]: Why did you change this? `deviceId` was correct.

File: pkg/server/server.go [hunk 4]
```diff
@@ -66,9 +66,11 @@ func (s *Server) createRouters() *negroni.Negroni {
...
+	r.HandleFunc("/data/{user}/sensor/{id}", dataController.GetByID).Methods("GET")
 	r.HandleFunc("/data", dataController.Create).Methods("POST")
+	r.HandleFunc("/data/{user}", dataController.DeleteAll).Methods("DELETE")
```
Comment (netoax) [inline pkg/server/server.go]: As I commented above, you should break this commit up so that the endpoints could be added separately.

File: pkg/interactor/data.go [hunk 5]
```diff
@@ -45,12 +46,12 @@ func (d *DataInteractor) Create(data Data) error {
 }
 
 func (d *DataInteractor) DeleteByDevice(deviceId string) (error error) {
-	err := dataStore.DeleteByDevice(deviceId)
+	err := dataStore.Delete(bson.M{"from": deviceId})
```
Comment (netoax) [inline pkg/interactor/data.go]: The interactor layer shouldn't know about database details (the `bson` format, in this case). Passing `deviceId` is completely fine.

**Review by netoax — CHANGES_REQUESTED** [review]

**Review by netoax — CHANGES_REQUESTED** [review]
_5b562fdec273558aa7d4d28ca2ace28ce5daad85_:
Please, follow [this](https://github.com/CESARBR/knot-babeltower/commit/da3412c36eed4e4d2ce9a79600e5ffd1460df34b) commit title and message pattern.
Something like: "data: controller: Add delete by id handler".

Also, in _ae11a71f72df53821d540cc722c71e26eec818c6_:
The tags are repeated and explain which component do you refactored: interactor, dataStore, etc...
About the message, I suggest you explain **_why_** you did this change, for example: to avoid repeated code.

**Comment by lcbm** [comment]
@Igoraamc please remember to rebase this with `migration_golang` :stuck_out_tongue:

File: pkg/interactor/data.go [hunk 6]
```diff
@@ -56,6 +56,22 @@ func (d *DataInteractor) Save(data Data) error {
...
+}
+
+func (d *DataInteractor) DeleteAll() error {
+	err := d.DataStore.Delete("")
+	if err != nil {
+		d.logger.Error(err)
+	}
+	return err
+}
+
```
Comment (lcbm) [inline pkg/interactor/data.go]: Error handling should be done right after checking for errors (inside the `if` statement). If no errors occur, we return `nil` instead. 

In this case, the functions are too small and it might not be trivial to understand the reason for the suggestion below. But imagine this:

1. `ControllerFoo` calls `InteractorMethod()`
2. `InteractorMethod()` then calls `functionFoo()` and `functionBar()`
3. `InteractorMethod()` handles returned errors for `functionFoo()` and `functionBar()`, as below:
```Go
func (i *interactor) InteractorMethod() {
  err := functionFoo()
  if err != nil {
    d.logger.Error(err)
  }
  err = functionBar()
  if err != nil {
    d.logger.Error(err)
  }
  return err
}
```
4. `ControllerFoo` (which called `InteractorMethod()`) would then receive the error from `functionBar()` instead of the primal error, which happened in `functionFoo()`.
5. `ControllerFoo` will then wrongfully handle the error, as well as we have added overhead (additional processing/complexity) unnecessarily, as we continued to run the subsequent instructions inside `InteractorMethod()` when we shouldn't have.

Right now, the code is quite simple and we wouldn't see many problems. However, as the codebase scales and we add functionalities further in the project, this could become a problem. To avoid this, I suggest we standardize error handling to be done within it's error-checking scope and return `nil` instead of `err` :point_down: 
```suggestion
func (d *DataInteractor) DeleteByDeviceID(deviceId string) error {
	err := d.DataStore.Delete(deviceId)
	if err != nil {
		d.logger.Error(err)
		return err
	}
	return nil
}

func (d *DataInteractor) DeleteAll() error {
	err := d.DataStore.Delete("")
	if err != nil {
		d.logger.Error(err)
		return err
	}
	return nil
}
```

File: pkg/controllers/data.go [hunk 7]
```diff
@@ -82,7 +84,27 @@ func (d *DataController) Save(w http.ResponseWriter, r *http.Request) {
 	}
 	respondWithJson(w, http.StatusCreated, thing)
 	defer r.Body.Close()
+}
+
+func (d *DataController) DeleteByDeviceID(w http.ResponseWriter, r *http.Request) {
```
Comment (lcbm) [inline pkg/controllers/data.go:105]: Shouldn't this, and other functions in here, return `error` so they can be handled appropriately? None of these operations are guaranteed to succeed and the caller should be aware when they don't, shouldn't they?

File: pkg/data/data_store.go [hunk 8]
```diff
@@ -56,3 +57,19 @@ func (ds *DataStore) Save(data Data) error {
...
+			ds.logger.Error(err)
+		}
+	} else {
+		_, err := ds.Database.C(collection).RemoveAll(nil)
+		if err != nil {
+			ds.logger.Error(err)
+		}
+	}
+	return err
+}
```
Comment (lcbm) [inline pkg/data/data_store.go]: Same as [here](https://github.com/CESARBR/knot-cloud-storage/pull/16#discussion_r407669994).

**Review by lcbm — CHANGES_REQUESTED** [review]
Awesome PR! :smile: 
Left some requests and comments, let me know when you take a look at them!

File: pkg/interactor/data.go (hunk 6 above)
Comment (Igoraamc) [reply pkg/interactor/data.go]: Done!

File: pkg/data/data_store.go (hunk 8 above)
Comment (Igoraamc) [reply pkg/data/data_store.go]: Done!

File: pkg/interactor/data.go (hunk 5 above)
Comment (Igoraamc) [reply pkg/interactor/data.go]: Done!

File: pkg/controllers/data.go (hunk 7 above)
Comment (Igoraamc) [reply pkg/controllers/data.go:105]: Hi @lcbm, as we talked in a call, the controller functions shouldn't return any error. This happens, because, the mux HandleFunc doesn't expect an error. For this reason, I'll be resolving this conversation.

**Comment by netoax** [comment]
@Igoraamc tip: our commits should follow the 50/72 rule. Please, take a look [here](https://chris.beams.io/posts/git-commit/#limit-50) and [here](https://chris.beams.io/posts/git-commit/#wrap-72).

File: pkg/interactor/data.go [hunk 9]
```diff
@@ -27,8 +27,9 @@ func (d *DataInteractor) GetAll(order, skip, take int, startDate, finishDate tim
 	data, err := d.DataStore.Get(selectOrder, skip, take, startDate, finishDate)
 	if err != nil {
 		d.logger.Error(err)
+		return nil, err
 	}
-	return data, err
+	return data, nil
```
Comment (netoax) [inline pkg/interactor/data.go]: Please, fix this in another commit or mention this change explicitly in the commit message.

File: pkg/interactor/data.go [hunk 10]
```diff
@@ -52,8 +53,27 @@ func (d *DataInteractor) Save(data Data) error {
 	err := d.DataStore.Save(data)
 	if err != nil {
 		d.logger.Error(err)
+		return err
+	}
+	return nil
```
Comment (netoax) [inline pkg/interactor/data.go]: Same here.

File: pkg/interactor/data.go [hunk 11]
```diff
@@ -52,8 +53,27 @@ func (d *DataInteractor) Save(data Data) error {
...
+}
+
+func (d *DataInteractor) DeleteByDeviceID(deviceId string) error {
```
Comment (netoax) [inline pkg/interactor/data.go]: Notice your changes here are related to the interactor layer instead of the database. So, please adjust the commit message to describe this.

File: go.sum [hunk 12]
```diff
@@ -1,6 +1,11 @@
 cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
 github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
+github.com/CESARBR/knot-cloud-storage v0.0.0-20200311161130-348492da29a9 h1:wma+1Z0K4H/L1j+5ZXJ2c58xS0lW6GQKlCPK38Ab/AM=
```
Comment (netoax) [inline go.sum]: These `go.sum` changes shouldn't be in this commit. I suggest you include them only when a new package is added based on the commit needs. Otherwise, changes that can't be placed in a specific commit should be added into a new one.

File: pkg/controllers/data.go [hunk 13]
```diff
@@ -82,7 +84,29 @@ func (d *DataController) Save(w http.ResponseWriter, r *http.Request) {
...
+	params := mux.Vars(r)
+	deviceId := params["deviceId"]
+
+	err := d.DataInteractor.DeleteByDeviceID(deviceId)
+	if err != nil {
+		respondWithError(w, http.StatusBadGateway, "Invalid information")
```
Comment (netoax) [inline pkg/controllers/data.go]: I think this error handling can be improved. The correct approach is to return the error according to the right protocol semantics as can be seen [here](https://tools.ietf.org/html/rfc7231#section-6.5). For example, when the client does the request with a wrong payload it's common to return the [`422 - Unprocessable entity`](https://evertpot.com/http/422-unprocessable-entity) status code.

File: pkg/controllers/data.go [hunk 14]
```diff
@@ -82,7 +84,29 @@ func (d *DataController) Save(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	respondWithJson(w, http.StatusOK, nil)
+}
+
+func (d *DataController) DeleteAll(w http.ResponseWriter, r *http.Request) {
+	err := d.DataInteractor.DeleteAll()
+	if err != nil {
+		respondWithError(w, http.StatusBadGateway, "Invalid information")
```
Comment (netoax) [inline pkg/controllers/data.go]: Same here about the error handling.

File: go.sum [hunk 15]
```diff
@@ -1,11 +1,6 @@
 cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
 github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
-github.com/CESARBR/knot-cloud-storage v0.0.0-20200311161130-348492da29a9 h1:wma+1Z0K4H/L1j+5ZXJ2c58xS0lW6GQKlCPK38Ab/AM=
-github.com/CESARBR/knot-cloud-storage v0.0.0-20200311161130-348492da29a9/go.mod h1:bbh6KJV1J+FJG9+oCcSubd9Sux5nhfD1L4FhCxvdEnw=
-github.com/Igoraamc/knot-cloud-storage v1.0.3 h1:YGxMrlzT/VEM6w7JfNAa03fWFjHNjK3nacyKpL7cwrM=
```
Comment (netoax) [inline go.sum]: This commit should contain only change related to the `mgo` package update.

File: go.mod [hunk 16]
```diff
@@ -9,6 +9,5 @@ require (
 	github.com/sirupsen/logrus v1.4.2
 	github.com/spf13/viper v1.6.1
 	gopkg.in/check.v1 v1.0.0-20200227125254-8fa46927fb4f // indirect
-	gopkg.in/mgo.v2 v2.0.0-20190816093944-a6b53ec6cb22
```
Comment (netoax) [inline go.mod:13]: Shouldn't this be done in the previous commit?

File: pkg/data/data_store.go [hunk 17]
```diff
@@ -5,15 +5,16 @@ import (
 
 	. "github.com/CESARBR/knot-cloud-storage/pkg/entities"
 	"github.com/CESARBR/knot-cloud-storage/pkg/logging"
-	"gopkg.in/mgo.v2"
-	"gopkg.in/mgo.v2/bson"
+	"github.com/globalsign/mgo"
```
Comment (netoax) [inline pkg/data/data_store.go:8]: It's important to notice this change should be grouped in the commit that updates `mgo` package in the `data/mongo.go` file too. The reason is that with only one commit the code remains cohesive and avoids breaking in an unexpected way.

File: pkg/data/data_store.go [hunk 18]
```diff
@@ -40,19 +41,38 @@ func (ds *DataStore) Get(order string, skip, take int, startDate, finishDate tim
...
 		ds.logger.Error(err)
+		return nil, err
 	}
 
 	if data == nil {
 		data = []Data{}
 	}
 
-	return data, err
+	return data, nil
```
Comment (netoax) [inline pkg/data/data_store.go:51]: Please, mention this additional fix in the commit message.

File: pkg/data/data_store.go [hunk 19]
```diff
@@ -40,19 +41,38 @@ func (ds *DataStore) Get(order string, skip, take int, startDate, finishDate tim
...
+
+func (ds *DataStore) Delete(deviceID string) error {
+	if deviceID != "" {
+		_, err := ds.Database.C(collection).RemoveAll(bson.M{"from": deviceID})
+		if err != nil {
+			ds.logger.Error(err)
+			return err
+		}
+	} else {
+		_, err := ds.Database.C(collection).RemoveAll(nil)
```
Comment (netoax) [inline pkg/data/data_store.go]: Couldn't this code be simplified to avoid this second call and those nested conditionals?

**Review by netoax — CHANGES_REQUESTED** [review]

File: pkg/server/server.go (hunk 3 above)
Comment (Igoraamc) [reply pkg/server/server.go]: Done!

File: pkg/interactor/data.go (hunk 9 above)
Comment (Igoraamc) [reply pkg/interactor/data.go]: Done!

File: pkg/interactor/data.go (hunk 10 above)
Comment (Igoraamc) [reply pkg/interactor/data.go]: Done!

File: pkg/interactor/data.go (hunk 11 above)
Comment (Igoraamc) [reply pkg/interactor/data.go]: Done!

File: pkg/data/data_store.go (hunk 18 above)
Comment (Igoraamc) [reply pkg/data/data_store.go:51]: Done!

File: pkg/data/data_store.go (hunk 17 above)
Comment (Igoraamc) [reply pkg/data/data_store.go:8]: Done!

File: pkg/controllers/data.go (hunk 13 above)
Comment (Igoraamc) [reply pkg/controllers/data.go]: Done!

File: pkg/controllers/data.go (hunk 14 above)
Comment (Igoraamc) [reply pkg/controllers/data.go]: Done!

File: go.sum (hunk 15 above)
Comment (Igoraamc) [reply go.sum]: Done!

File: go.mod (hunk 16 above)
Comment (Igoraamc) [reply go.mod:13]: Done!

File: pkg/data/data_store.go (hunk 19 above)
Comment (lcbm) [reply pkg/data/data_store.go]: Something along these lines should work (not sure, tho) :point_down: 

```Go
func (ds *DataStore) Delete(deviceID string) error {
	if deviceID == "" {
		return ds.removeAll(nil)
	}
	return ds.removeAll(bson.M{"from": deviceID})
}

func (ds *DataStore) removeAll(query <type>) error {
	_, err := ds.Database.C(collection).RemoveAll(query)
	if err != nil {
		ds.logger.Error(err)
		return err
	}
	return nil
}
```

File: pkg/interactor/data.go (hunk 2 above)
Comment (esdrasjnr) [reply pkg/interactor/data.go]: Agree with @joaoaneto! As these files will be in same package, you can just move the code around those files. You can also create a file called `interactor.go` to hold the base code, as the `DataInteractor` struct definition and `NewDataInteractor` constructor method. This is what we do in [`babeltower` interactor package](https://github.com/CESARBR/knot-babeltower/tree/master/pkg/thing/interactors).

Let us know if you have some questions or suggestions! :100:

File: pkg/data/data_store.go (hunk 19 above)
Comment (esdrasjnr) [reply pkg/data/data_store.go]: Good suggestion @lcbm!

File: pkg/interactor/data.go [hunk 20]
```diff
@@ -87,6 +88,25 @@ func (d *DataInteractor) Save(token, id string, data []entities.Payload, timesta
...
 	err := d.UsersService.Authenticate(token)
 	if err != nil {
+		d.logger.Error(err)
+		return err
+	}
+	return nil
+}
+
+func (d *DataInteractor) DeleteByDeviceID(deviceId string) error {
+	err := d.DataStore.Delete(deviceId)
```
Comment (esdrasjnr) [inline pkg/interactor/data.go]: Notice that the `Delete` method doesn't exists in this commit yet. I suggest you to replace it by a `TODO` code comment and add this method call in the commit you create that method. This is important to avoid creating inconsistent commits :smiley:

File: pkg/interactor/data.go [hunk 21]
```diff
@@ -87,6 +88,25 @@ func (d *DataInteractor) Save(token, id string, data []entities.Payload, timesta
...
+	if err != nil {
+		d.logger.Error(err)
+		return err
+	}
+	return nil
+}
+
+func (d *DataInteractor) DeleteAll() error {
+	err := d.DataStore.Delete("")
```
Comment (esdrasjnr) [inline pkg/interactor/data.go]: Same here! :+1:

File: pkg/controllers/data.go [hunk 22]
```diff
@@ -39,9 +39,11 @@ func (d *DataController) respondWithJson(w http.ResponseWriter, code int, payloa
...
 	w.WriteHeader(code)
-	_, err := w.Write(response)
-	if err != nil {
-		d.logger.Error(err)
+	if payload != nil {
+		_, err := w.Write(response)
+		if err != nil {
+			d.logger.Error(err)
+		}
 	}
```
Comment (esdrasjnr) [inline pkg/controllers/data.go:47]: I think that an `else` statement to log that `payload` is empty would be good here, what do you think @Igoraamc?

**Review by esdrasjnr — CHANGES_REQUESTED** [review]
Excellent patch @Igoraamc!!
I made a few comments and suggestions, let me know if you have any questions or suggestions!
:+1: :100:

File: pkg/interactor/data.go (hunk 2 above)
Comment (netoax) [reply pkg/interactor/data.go]: As discussed in the `slack` channel, I'd prefer you add the delete in another file `delete_data.go` but without following the babeltower pattern.

File: pkg/data/data_store.go (hunk 19 above)
Comment (lcbm) [reply pkg/data/data_store.go]: This conversation is still pending :stuck_out_tongue:

**Review by lcbm — CHANGES_REQUESTED** [review]
Awesome update on the patch! :100: 

In respect to what I've commented, there's a pending conversation from the last review as well as a new observation regarding the commit message title: 

- _Commit titles and messages should be written in imperative mode._

Here's a [modified] [classic but good example](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html): 

>Capitalized, short (50 chars or less) summary
>
>More detailed explanatory text, if necessary.  Wrap it to about 72
characters or so. The blank line separating the summary from the 
body is critical (unless you omit the body entirely).
>
>Write your commit message in the imperative: "Fix bug" and not "Fixed bug"
or "Fixes bug."  This convention matches up with commit messages generated
by commands like `git merge` and `git revert`.

**Review by esdrasjnr — APPROVED** [review]
Awesome, LGTM :smiley:

File: pkg/data/data_store.go (hunk 19 above)
Comment (netoax) [reply pkg/data/data_store.go]: @Igoraamc I think the @lcbm suggestion is a good one. :)

**Review by netoax — CHANGES_REQUESTED** [review]
LGTM!
Just take a look at the [comment](https://github.com/CESARBR/knot-cloud-storage/pull/16#pullrequestreview-394027899) @lcbm did with respect to commit messages.

**Review by netoax — APPROVED** [review]
